R/quantile_twas_weight.R

```diff
@@ -0,0 +1,485 @@
+        stop("Invalid method. Choose 'fdr' or 'qvalue'.")
+    }
+
+    sig_SNP_threshold = which(adjusted_pvalues < threshold)
+    sig_SNP_top_count = order(adjusted_pvalues)[1:top_count]
+    sig_SNP_top_percent = order(adjusted_pvalues)[1:max(1, round(length(adjusted_pvalues) * top_percent / 100))]
+
+    sig.SNPs_names = colnames(X)[sig_SNP_threshold]
+    sig.SNPs_names_top_count = colnames(X)[sig_SNP_top_count]
+    sig.SNPs_names_top_percent = colnames(X)[sig_SNP_top_percent]
+    phenotype_id = colnames(y)[1]
+
+    df_result = data.frame(
+        phenotype_id = phenotype_id,
+        variant_id = colnames(X),
+        p_qr = pvec
+    )
+
+    # Add quantile-specific p-values
+    for (tau in tau.list) {
+        df_result[[paste0("p_qr_", tau)]] = quantile.pvalue[, paste0("p_qr_", tau)]
+    }
+
+    # Add overall q-value
+    df_result[[method_col_name]] = adjusted_pvalues
+
+    # Add quantile-specific q-values
+    for (tau in tau.list) {
+        df_result[[paste0(method_col_name, "_", tau)]] = quantile_adjusted_pvalues[, paste0("p_qr_", tau)]
+    }
+
+    # Add quantile-specific z-scores
+    for (tau in tau.list) {
+        df_result[[paste0("zscore_qr_", tau)]] = quantile.zscore[, paste0("zscore_qr_", tau)]
+    }
+
+    # Split variant_id and reorder columns
+    df_result <- df_result %>%
+        separate(variant_id, into = c("chr", "pos", "ref", "alt"), sep = "[:_]", remove = FALSE) %>%
+        mutate(chr = as.numeric(gsub("chr", "", chr)), pos = as.numeric(pos))
+
+    # Define the column order
+    col_order <- c("chr", "pos", "ref", "alt", "phenotype_id", "variant_id", "p_qr")
+    col_order <- c(col_order, paste0("p_qr_", tau.list))
+    col_order <- c(col_order, method_col_name)
+    col_order <- c(col_order, paste0(method_col_name, "_", tau.list))
+    col_order <- c(col_order, paste0("zscore_qr_", tau.list))
+
+    # Reorder the columns
+    df_result <- df_result %>% select(all_of(col_order))
+
+    return(list(df_result = df_result,
+                tau_list = tau.list, 
+                quantile_pvalue = quantile.pvalue, 
+                quantile_zscore = quantile.zscore,
+                pvec = pvec, 
+                adjusted_pvalues = adjusted_pvalues, 
+                sig_SNP_threshold = sig_SNP_threshold, 
+                sig.SNPs_names = sig.SNPs_names,
+                sig_SNP_top_count = sig_SNP_top_count, 
+                sig_SNP_top_percent = sig_SNP_top_percent))
+}
+
+#' Perform Clumping and Pruning
+#' @param X Matrix of genotypes
+#' @param qr_results Results from QR_screen
+#' @param maf_list List of minor allele frequencies (optional)
+#' @param ld_clump_r2 R-squared threshold for initial LD clumping based on pvalue
+#' @param final_clump_r2 R-squared threshold for final LD clumping based on MAF
+#' @return A list containing final SNPs and clumped SNPs
+#' @export
+perform_clumping_and_pruning <- function(X, qr_results, maf_list = NULL, ld_clump_r2 = 0.2, final_clump_r2 = 0.8) {
```

Comment:
could you change the function name to `multicontext_ld_clumping` to better reflect what it does?